### ISSUE TITLE
Add a IRR as-set/route-set field resp. irr_as_set6 for IPv6

### ISSUE THREAD

**Issue by arnoldnipper** [peeringdb/peeringdb#605]
Although IPv6 is around for two decades some networks still have different routing policies for IPv4 and IPv6. To support this a new field "IRR as-set/route-set IPv6" resp. irr_as_set6 would be needed. This would also allow for a clear structure as this field only can have a single value then.

Relates to #26 #151 and #260

**Comment by shane-kerr** [comment]
In RPSL you can declare as many AS-SET objects as you want, which allows networks with different routing policies to have an AS-SET for IPv4 and an AS-SET for IPv6, if that matches their deployment. Is this not possible with the PeeringDB mapping?

**Comment by job** [comment]
RPSL allows for very advanced stuff, but nobody uses those features. In
practice virtually everyone assumes one as-set

**Comment by arnoldnipper** [comment]
While we all agree that in the end, everyone will only use **one** `irr_as_set`, reality shows that ~40 networks have specified more than one `irr_as_set`. more of half obviously wanting to express a different policy for IPv4 and IPv6.

@shane-kerr pointed out that a network may have as many AS-SET objects as they want. However, IMHO this clearly is not mappable to PeeringDB, e.g. to have a different AS-SET at any IX.

To move this issue forward and to implement `irr_as_set` validation, I see two options.

 * option1: add `irr_as_set6` (which basically is this issue)

 * option2: leave everything as is and discuss the rest in #151 

@peeringdb/pc, please make up your mind

**Comment by grizz** [comment]
> @shane-kerr pointed out that a network may have as many AS-SET objects as they want. However, IMHO this clearly is not mappable to PeeringDB, e.g. to have a different AS-SET at any IX.

That would be possible if we added an AS-SET (and AS-SET6) field to the `netixlan` data.

I would be in favor of adding `irr_as_set6` for sure, but would like to tackle how we're splitting and validating multiple AS-SETs in #151 first, as we may learn some things in doing the actual implementation. Once #151 is done, we can discuss how to best implement this.

**Comment by arnoldnipper** [comment]
> I would be in favor of adding `irr_as_set6` for sure, but would like to tackle how we're splitting and validating multiple AS-SETs in #151 first, as we may learn some things in doing the actual implementation. Once #151 is done, we can discuss how to best implement this.

For me, it's just the other way round, @grizz 💅 ... if we decide to have `irr_as_set6` then we would only allow at most one as-set/route-set. If we don't we would allow up to two. Either a global one or one for IPv4 and one for IPv6. We then would leave it to the users to find out which is which. 

If we have {,2} then I don't care what the separator is. Ideally ", " or simply ",".

**Comment by arnoldnipper** [comment]
@grizz would you please explain why you like to tackle how we're splitting and validating multiple AS-SETs in #151 first? For me, if we add `irr_as_set6`, both `irr_as_set` and `irr_as_set6` only would have one as-set/route set as value. And for me, this seems the only way to go as multiple as-sets/route-sets in `irr_as_set` would eventually mean that the peer has to build the closure of all of them.

**Comment by grizz** [comment]
@arnoldnipper a good reason to do #151 first is that if we split and validate, I don't believe anyone will argue, as it's just cleaning up data.

Adding fields and limiting the number of records are going to end up being a bigger discussion including disagreements, as we've seen. Yet doing any of that still requires having clean data, so not only is #151 a prerequisite it's something we could knock out fairly easily to improve data quality, and then we can all pedant about how the records should look. :)

Once the data is clean for people to start using it, we can see what the exact needs are.

**Comment by job** [comment]
Without voting one way or another, I have some questions:
 
How do we migrate from `irr_as_set` to `irr_as_set + irr_as_set6`?

Do we start out with `irr_as_set6` being empty? Do we expect data consumers to probe whether `irr_as_set6` is filled in, and if not, fall back to `irr_as_set` for the IPv6 generation?

What process do we envision should `irr_as_set6` exist?

**Comment by arnoldnipper** [comment]
> Without voting one way or another, I have some questions:

 * currently we assume that an `irr_as_set` may hold information for both IPv4 and IPv6, if there is more than one as-set. If we decide to have `irr_as_set` and  `irr_as_set6` then we should also decide that there is at most *one* as-set for each of them

 * if we agree on that we would inform our users about what I mentioned above and what the meaning, as well as proposed procedure for dealing with the new field, is

 * we add `irr_as_set6` with the next release leaving the field empty for all networks

 * meanwhile AC would contact all networks having an `irr_as_set` with more than one as-set and work with them on how to split the field.

Would that work?

**Comment by arnoldnipper** [comment]
Now that IRR checking is implemented (we still have to do a cleanup!), let's address this now. @peeringdb/pc. Please make up your mind!

**Comment by grizz** [comment]
If we're going to split them, it might make more sense to add `irr_as_set4` and `irr_as_set6` and then turn `irr_as_set` into a field that combines the two on read.

**Comment by arnoldnipper** [comment]
And already announce that `irr_as_set` will be phased out somewhen (e.g v3)?

**Comment by grizz** [comment]
If we get rid of it, we need to bump to version 3.

I'd think we make it writable but it writes to irr_as_set4 so no code would have to change.

Then, like you said, we can get rid of it in v3.

**Comment by job** [comment]
I'm not sure we should continue down this path: we're adding more and more routing policy elements to PeeringDB without first doing a thorough study of what is fully needed in this space. We don't want to end up re-inventing the RPSL wheel inside PeeringDB, that is work that should be taken to the IETF.

An issue I have with this split of AS-SET objects is that for years we've as community tried to advocate that if IPv6 is to make a fighting chance, people need to dual-stack *everything*. In the transit world there is so much IPv6 deployment already (what real transit provider doesn't support IPv6?) that the differences between the "IPv4" and "IPv6" AS-SETs are minimal at best, and operationally mostly irrelevant (as has been proven by the fact that PeeringDB survived for years without the AFI separation). If there is IPv4 BGP between two networks, I assume there is a good chance IPv6 BGP peering already exists (but perhaps not visible to me), or can be set up quickly.

Indeed there are peering disputes in the IPv6 default-free zone, but those are *peering* disputes, which doesn't really impact the customer cone of any party involved. And this field is expected to represent the customer cone of what you'd announce to a peer. So topological differences between the IPv4 and IPv6 DFZ to me is not an argument for AFI separation in the PeeringDB context.

If I talk to fellow operators, I recommend them to create 1 AS-SET and put their downstream customers in that. When I generate an IPv4 BGP prefix list filter I just query the IRR for `route:` objects, and for IPv6 you query via an inverse lookup for `route6:` objects. Asking people to maintain (often manually) multiple IRR objects is  raising the bar in a way that I fear might not outweigh the 'purity' argument that may apply in this ticket.

If PeeringDB offers AFI separated AS-SET fields, people might become inspired to start using more than one AS-SET, and this might cause stale data in the IRR ecosystem later on.

In summary: what we offer today is "good enough" for today's operational best current practises. Innovation is needed in this space, but we should do that in IETF, adding more fields to PDB just clutters the UI for no real benefit. (in my perception, speaking as isp + route server operator, author of irr automation tools)

**Comment by arnoldnipper** [comment]
> If we get rid of it, we need to bump to version 3.
> 
> I'd think we make it writable but it writes to irr_as_set4 so no code would have to change.
> 
> Then, like you said, we can get rid of it in v3.

+1 sounds like a good plan. However, that needs good explanation.

**Comment by durkovic** [comment]
Just a real-world example:

AS6939, connected to 230+ IXPs, uses: 
AS-HURRICANE (19260 ASs)
AS-HURRICANEv6 (10641 ASs, many of them unique for IPv6)  

Largest IXPs connect 100+ networks with different IPv4/IPv6 macros, and reflect this in their route-server AS-SETs. 

However, there's no way to reasonably document this in PeeringDB.

**Comment by job** [comment]
If we look at the world, let's look at the real real world :-)

<img src="http://instituut.net/~job/screenshots/78110b5f9ba660b1.png"/>

**Comment by job** [comment]
I am intimately familiar with these objects and what they mean.
 
```
hanna:he job$ bgpq4 -6 AS-HURRICANEv6 | sort > AS-HURRICANEv6
hanna:he job$ bgpq4 -6 AS-HURRICANE | sort > AS-HURRICANE
hanna:he job$ comm -1 -2 AS-HURRICANE AS-HURRICANEv6 | wc -l
140084
hanna:he job$ wc -l AS-HURRICANE*
  164761 AS-HURRICANE
  247627 AS-HURRICANEv6
```

What we can learn from the above is that there is a ton of overlap between `AS-HURRICANE` and `AS-HURRICANEv6`. This is an artifact of how the `AS-SET` is generated in the first place, and how the IPv4 and IPv6 DFZ are structured.

With `AS-SETs` of this size and scope, other protection mechanisms such as RPKI ROV and Peerlock come into play to protect yourself. Differentiating the AS-SETs on AFI doesn't matter that much in this case, the objects are both far wider than reality anyway.

Can we close this ticket and move on to other work?

**Comment by durkovic** [comment]
The overlap is very well justified - it's all the dual-stack networks. However, there are still many IPv4-only peers and specifically in the HE case, many IPv6-only peers. 

Networks and IXPs configure IPv4 and IPv6 separately, there are separate IP addresses, separate max-prefix configurations and separate filters. All these settings can be exported via IX-F json schema, where as_macro was deliberately put under both ipv4 & ipv6 sections and not under member section.

Whether you like it or not, it's used in the real world:

AS-DECIX  and  AS-DECIX-V6
AS-AMS-IX-RS-SETS  and  AS-AMS-IX-RS-SETS-V6

PeeringDB still lacks reasonable way to document it, both of the above IXPs only use the first as-macro here. As a result, all IPv6-only ASs are clearly missing.

**Comment by job** [comment]
You are missing that you issue a different command (`!g` vs `!6`) to the `IRRd` server when you expand an IPv6 vs an IPv4 set. If a network is IPv6-only, or IPv4-only, no prefixes (or ASNs) would contribute to the generated filters if you use `bgpq3` or `bgpq4`.

If a network is single-stack: it does not pollute the filters of the other AFI. If the network is dual-stack, from a business operations perspective there is a relation between the peer-ASN and the potential downstream customers. Especially in the cases of route servers peers can come & go any time so its really just about /potential/ downstreams.

I'm saying that maintaining separate AS-SETs for IPv4 and IPv6 is not a productive practise, it doesn't serves a real purpose other than additional objects to maintain. So now separating them in PeeringDB is a lot of work for a lot of people and I don't see an upside. What problem are you /really/ solving? Can you show me an example based on how in your network something got accepted which you think should not have been accepted based on the 'real world' example you brought up?

Another thing to consider: PeeringDB has not had AFI separated fields for 15 years, and it was not a problem in practise. Otherwise, how would it have worked so well so far?

I think we should promote the idea of using a single AS-SET because the AFI separation happens deeper down in the stack following whether IPv4 or IPv6 config output is requested. So if users fill multiple AS-SETs in, peeringdb's webinterface can throw a warning: "Are you use you want to use multiple AS-SETs? The BCP is to use a single AS-SET"

**Comment by nickhilliard** [comment]
couple of observations on this:  

- as-sets are a component of a grossly over-simplified model of real-world routing policies
- ideally, prefix filtering needs to happen on the basis of prefix + origin as + as-path. From this point of view, the entire concept of an as-set is only ever going to be a rough approximation of what's actually required to implement a set of {prefix,origin as,as-path} n-tuples.  Note that hardly any routing stack grammars can even handle this.
- trivially-scaled networks usually have congruency as-set4 and as-set6s
- anything larger than a trivial network are likely to have different sets of these n-tuples at different locations.  The larger the network, the more likely it is that this will happen.
- designing tools around consumption of AS-HURRICANE and AS-HURRICANE6 is not a good idea, other than using the as-sets as a worst-case scenario for scaling / testing your tools.  This is particularly the case given how these as sets are generated by HE.
- from an endpoint perspective, adding bells and whistles on the peeringdb export api creates an equivalent obligation for data consumers to tool this up

I get why this seems like a good idea and am not going to claim that it is or isn't. The issue is that there needs to be a more broad understanding that 1) this isn't as useful for good quality policy routing management as we'd prefer to believe and 2) why this is the case, i.e. the irrdb model is fundamentally oversimplified and putting in a tweak for a specific data point masks deeper underlying problems of the data model at the cost of complicating things in a way that is, tbh, more than slightly hacky.

**Comment by durkovic** [comment]
@job As an IXP, we initially also supported only a single AS-SET per IX member. However, this created confusion and repeated requests from our members to support one AS-SET per AFI.
As mentioned above, large IXPs have more than 100 members, which maintain separate AS-SETs per AFI, and, as a consequence, maintain separate AS-SETs as well. 

The only way to put such setup into PeeringDB is to store both IPv4 and IPv6 AS-SETs in a single field, which is at best a hack and probably causes parsing problems, since both DECIX and AMSIX decided to only put their IPv4 AS-SETs into PeeringDB. When someone compiles filters according to this, the filters will clearly lack all prefixes of IPv6-only ASs. Although this might seem a minor problem now, number of IPv6-only ASs will certainly increase over time.

IMHO, it's much cleaner to have 2 fields, both restricted to single AS-SET. This way, IXPs can properly propagate IPv4-only and IPv6-only members - please have in mind, that IPv4-only members can't establish peering with IPv6-only members, so the unified AS-SET is clearly misleading in IXP context.

Yes, the current setup was working for 15 years, but many networks use AS-SETs from PeeringDB  only when IRR does not provide anything useful.

**Comment by job** [comment]
It is possible we won't come to an agreement.

**Comment by arnoldnipper** [comment]
> It is possible we won't come to an agreement.

Leaving as it is is the worst we can do IMO.

Assuming that there is a per AFI `irr_as_set` do we all agree that then the value would be a single AS-SET?

**Comment by job** [comment]
> > It is possible we won't come to an agreement.
> 
> Leaving as it is is the worst we can do IMO.

Actually it is not: most generator software (arouteserver, ixp manager?) takes the single or multiple values that are in the IRR field and passes those on as query arguments to the configuration generator. It is valid in RPSL to use multiple values (whitespace separated) in this context. Then depending on the task at hand: provisioning an IPv4 or an IPv6 filter a different filter will come out. This assumes the industry intends to continue with a dual-stack strategy (as we've been doing the last years). This aligns with POLA.
 
> Assuming that there is a per AFI `irr_as_set` do we all agree that then the value would be a single AS-SET?

```
¯\_(ツ)_/¯
```

This is the trouble: we don't have a formal definition what the data means, it is nothing more than a hint. If people need more granularity, they really need to negotiate directly with the peering network and agree to use RPSL in the IRR, or different out-of-band mechanism to discover what AS-SET should be used for what EBGP peering point.

I think it is out of scope for PeeringDB to attempt to redefine a routing language, this is work that has to happen in IETF.

I hate arguments based on authority, but please note this thread observed resistance to the idea from me (engineer on IRRd, bgpq3, bgqp4, arouteserver, multiple IXPs, NTT), and Nick (maintainer of IRRToolset, RPSL expert) also noted considerable complications that need to be accounted for.

The current situation now that we have syntax validation is quite good and can persist for many years: it motivates people to unify their IPv4 and IPv6 peering policies and filtering mechanism. If we want to promote IPv6 as industry, we should probably focus on making the difference between IPv4 and IPv6 less hard.

**Comment by nickhilliard** [comment]
> Leaving as it is is the worst we can do IMO.

not really: the underlying data model is wrong. You can't fix it with a tweak.

> Assuming that there is a per AFI `irr_as_set` do we all agree that then the value would be a single AS-SET?

Not sure what the development model is here, but I don't see consensus that a per AFI `irr_as_set` is either a good idea or worth implementing.

**Comment by grizz** [comment]
Upon evaluation, I think it's clear that the best case is leaving this as is.

When I stopped and thought about it, I came to the same conclusion as to why I've only ever used a single as-set for both AFIs -- when you use them IRL, you pass the prefixes to something that does a look up and creates a prefix list. The thing that makes the prefix list will separate by AFI if needed.

We aren't going to verify that the AS-SET entered in PDB contains only prefixes from the address family for that field, so we shouldn't be showing otherwise.

I agree with Job and Nick, it's a flawed data model, but it's being used in it's current form.

-1 from me.

@peeringdb/pc please vote so we can get this issue off our plate

**Comment by arnoldnipper** [comment]
As there is no consensus and the status quo works for almost everyone

**Comment by as44980** [comment]
> As there is no consensus and the status quo works for almost everyone

While I dont see consensous, I don not see any solution to the original issues raised... was there a vote as suggested by @grizz ?
